sdk/include/opentelemetry/sdk/common/http_client.h

```diff
@@ -0,0 +1,149 @@
+#pragma once
+
+#include <chrono>
+#include <string>
+#include <vector>
+#include "opentelemetry/nostd/function_ref.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/version.h"
+
+/*
+ Usage Example
+
+ struct SimpleReponseHandler: public ResponseHandler {
+      void OnResponse(Response& res) noexcept override
+      {
+           if (res.IsSuccess()) {
+            res.GetNextHeader([](nostd::string_view name, std::string value) -> bool {
+                std::cout << "Header Name:" << name << " Header Value:"<< value ;
+                return true;
+            });
+            .. process response body
+           }
+      }
+
+      void OnError(nostd::string_view err) noexcept override
+      {
+          std::cout << " Error:" << err;
+      }
+  };
+
+  SessionManager sessionManager; // implementer can provide singleton implementation for it
+  auto session = sessionManager.createSession("localhost", 8000);
+  auto request = session->CreateRequest();
+  request->AddHeader(..);
+  SimpleResponseHandler res_handler;
+  session->SendRequest(res_handler);
+  session->FinishSession() // optionally in the end
+  ...shutdown
+  sessionManager.FinishAllSessions()
+*/
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace common
+{
+namespace http
+{
+
+enum class Method
+{
+  Get,
+  Post,
+  Put,
+  Options,
+  Head,
+  Patch,
+  Delete
+};
+
+enum class SessionState
+{
+  Created,  // session object is created
```

Comment:
Basically we don't currently cover negative scenarios with this API. Real-world scenarios would include connection failed, either timeout or invalid certificate. That is covering error conditions even before we can report HTTP status code (because there is no HTTP response code for connection failures, yet the code has to handle these).

Reply:
+1